### PR TITLE
test(quantic): added  playwright tests and unit test for quantic generated answer component

### PR DESCRIPTION
**Pull request by mmitiche**
## [SFINT-5789](https://coveord.atlassian.net/browse/SFINT-5789)

- Playwright E2E tests added for the Quantic Generated Answer component.
- Jest unit tests added for the Quantic Generated Answer component.

## Unit tests:

Screenshot 2024-12-19 at 3 14 07 PM

## E2E tests

Screenshot 2024-12-19 at 3 21 04 PM

[SFINT-5789]: https://coveord.atlassian.net/browse/SFINT-5789?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/quanticGeneratedAnswer.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            documented: 'unknown',
+            readable: 'yes',
+            documentUrl: exampleDocumentUrl,
+            details: exampleDetails,
+            helpful: true,
+          });
+        await generatedAnswer.clickSubmitFeedbackButton();
+        await feedbackAnalyticRequestPromise;
+
+        await page.waitForTimeout(2000);
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/quanticGeneratedAnswer.e2e.ts:62]:
```suggestion
```

Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/quanticGeneratedAnswer.e2e.ts:62]:
```suggestion
```

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+  constructor(
+    public page: Page,
+    public streamId: string
+  ) {
+    this.page = page;
+    this.streamId = streamId;
+  }
+
+  get likeButton(): Locator {
+    return this.page.locator('[data-cy="feedback__like-button"]');
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:19]: If possible try to use the `getByRole` from Playwright, it is what's closer to "accessibility" testing we have for now.

Combine them with the `.getByRole().filter` to make specific Locators.

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    );
+  }
+
+  get toggleButton(): Locator {
+    return this.page.locator('[data-test="generated-answer__toggle-button"]');
+  }
+
+  feedbackOption(value: string, index: number): Locator {
+    return this.page
+      .locator('.feedback-modal-qna__question input', {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:38]: This can be made better with something like this I'm sure:

`page.getByRole('input').filter({hasText : ???})`

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    return this.page.locator(
+      '[data-test="generated-answer__citations"] .citation__link'
+    );
+  }
+
+  async hoverOverCitation(index: number): Promise<void> {
+    // waiting 500ms to allow the component to render completely, cause any re-rendering abort the hover action.
+    await this.page.waitForTimeout(500);
+    await this.citationLink.waitFor({state: 'visible'});
+    await this.citationLink.nth(index).hover();
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:82]: This already waits for "actionability" with Playwright, not sure if you need the `state: visible` ?

Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:82]: This already waits for "actionability" with Playwright, are you sure you need the above waitFor?

> * **Details**
   *
   * This method hovers over the element by performing the following steps:
   * 1. Wait for [actionability](https://playwright.dev/docs/actionability) checks on the element, unless
   *    [`force`](https://playwright.dev/docs/api/class-locator#locator-hover-option-force) option is set.
   * 1. Scroll the element into view if needed.

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+
+  async typeInFeedbackDetailsInput(text: string): Promise<void> {
+    await this.feedbackDetailsInput.fill(text);
+  }
+
+  async clickSubmitFeedbackButton(): Promise<void> {
+    await this.submitFeedbackButton.click();
+  }
+
+  async fillFeedbackForm(answers: Record<string, string>): Promise<void> {
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:103]: This is kinda heavily relying on the fact that the text of the questions won't change right?
Wouldn't it be better to either use their "tags"? (I think we have the tag "relevant" "hallucination" etc. on each questions.)
Or maybe simply accept just a list of answers and fill them in order?

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+  async mockStreamResponse(
+    body: Array<{payloadType: string; payload: string; finishReason?: string}>
+  ) {
+    await this.page.route(
+      `**/machinelearning/streaming/${this.streamId}`,
+      (route) => {
+        let bodyText = '';
+        body.forEach((data) => {
+          bodyText += `data: ${JSON.stringify(data)} \n\n`;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:282]: There you go! GG

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/templates/generatedAnswer.html
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/templates/generatedAnswer.html:1]: Format with Prettier please, there will be much less changes :)

**Review by erocheleau — COMMENTED** [review]
Small suggestions but pretty much there!

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts (hunk 4 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:82]: Indeed that's true you are right,
I added this line to try to solve the issue I got where a component re-render  discarded the hover, a problem that later I solved with lines 79 and 80 but forgot to delete the unnecessary line 81 👍

**Review by erocheleau — COMMENTED** [review]
Guess what @mmitiche, I found another cool thing Playwright offers :D 

https://playwright.dev/docs/locators#locate-by-test-id

If you set `data-testId` on your HTML elements, you can simply do `getByTestId('')` to locate an element. instead of having to do '[data-test="generated-answer__copy-to-clipboard"]' for example

File: packages/quantic/cypress/e2e/default-2/generatedAnswer/generated-answer-selectors.ts [hunk 7]
```diff
@@ -36,7 +36,7 @@ export const GeneratedAnswerSelectors: GeneratedAnswerSelector = {
   get: () => cy.get(generatedAnswerComponent),
 
   generatedAnswerCard: () =>
-    GeneratedAnswerSelectors.get().find('[data-cy="generated-answer__card"]'),
+    GeneratedAnswerSelectors.get().find('[data-test="generated-answer__card"]'),
```
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/cypress/e2e/default-2/generatedAnswer/generated-answer-selectors.ts:39]: Here I started moving gradually to `data-test` because there are some selectors that already exist in the Cypress tests and that I wanted to reuse in the new unit tests, so instead of creating more places that use data-cy that we know will be removed soon, I thought about gradually starting to switch to data-test, and reduce the places where we use data-cy.

and of course when delete our cypress setup, we remove every remaining occurrence of data-cy.

To summarize, I wanna avoid having data-cy in our unit tests as this increase the volume of work needed when cleaning out all cypress and data-cy.

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js [hunk 8]
```diff
@@ -51,6 +101,18 @@ const exampleEngine = {
   id: 'dummy engine',
 };
 let isInitialized = false;
+// let mockAnswerHeight = 300;
```
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js:104]: just a forgotten comment, deleted.

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js [hunk 9]
```diff
@@ -72,19 +143,125 @@ function cleanup() {
...
+
+      const initializationError = element.shadowRoot.querySelector(
+        selectors.initializationError
+      );
+
+      expect(initializationError).not.toBeNull();
+    });
+  });
+
+  describe('when an RGA retryable error occurs', () => {
```
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js:167]: I think `an RGA` is the correct way 🤨

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts (hunk 5 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts:103]: Addressed 👌

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts (hunk 2 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/e2e/pageObject.ts]: improved 👍

File: packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js:166]: wow wtf you're right I just learned that today for acronyms it's what the letter "sounds" like? lol what a bogus rule...

because it's "arr-gee-ay" we should use "an" :joy:

**Review by erocheleau — APPROVED** [review]
Made some modifications to use the `getByTestId` selector from playwright mainly.
Updated one selector to use `getByRole` too

File: packages/quantic/cypress/e2e/default-2/generatedAnswer/generated-answer-selectors.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/cypress/e2e/default-2/generatedAnswer/generated-answer-selectors.ts]: I updated your pr to `data-testid` for future playwright.
